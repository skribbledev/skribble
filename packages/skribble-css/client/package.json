{
  "exports": {
    ".": {
      "import": "./../dist/client.esm.js",
      "require": "./../dist/client.cjs.js",
      "types": "./../dist/client.d.ts"
    },
    "./index.js": {
      "import": "./../dist/client.esm.js",
      "require": "./../dist/client.cjs.js",
      "types": "./../dist/client.d.ts"
    }
  },
  "main": "../dist/client.cjs.js",
  "module": "../dist/client.esm.js",
  "types": "../dist/client.d.ts"
}
