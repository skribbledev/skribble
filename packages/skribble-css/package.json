{
  "name": "skribble-css",
  "version": "0.0.0",
  "description": "A better way to work with typed css",
  "keywords": [
    "skribble",
    "napi",
    "rust",
    "skribble css",
    "css"
  ],
  "repository": "https://github.com/skribbledev/skribble/tree/master/packages/skribble-css",
  "license": "MIT",
  "contributors": [],
  "sideEffects": false,
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
      "types": "./dist/index.d.ts"
    },
    "./client": {
      "import": "./dist/client.esm.js",
      "require": "./dist/client.cjs.js",
      "types": "./dist/client.d.ts"
    },
    "./client.js": {
      "import": "./dist/client.esm.js",
      "require": "./dist/client.cjs.js",
      "types": "./dist/client.d.ts"
    },
    "./index.js": {
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
      "types": "./dist/index.d.ts"
    },
    "./package.json": "./package.json",
    "./types/*": "./dist/*.d.ts"
  },
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "types": "./dist/index.d.ts",
  "files": [
    "client",
    "dist"
  ],
  "dependencies": {
    "@node-rs/helper": "^1.3.3",
    "@skribble-css/client": "^0.0.0",
    "tiny-invariant": "^1.2.0",
    "ts-extras": "^0.10.2"
  },
  "devDependencies": {
    "@types/node": "^17.0.30"
  },
  "peerDependencies": {
    "@types/node": "*"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    }
  },
  "optionalDependencies": {
    "skribble-css-android-arm64": "^0.0.0",
    "skribble-css-android-eabi": "^0.0.0",
    "skribble-css-darwin-arm64": "^0.0.0",
    "skribble-css-darwin-x64": "^0.0.0",
    "skribble-css-freebsd-x64": "^0.0.0",
    "skribble-css-linux-arm-gnueabihf": "^0.0.0",
    "skribble-css-linux-arm64-gnu": "^0.0.0",
    "skribble-css-linux-arm64-musl": "^0.0.0",
    "skribble-css-linux-x64-gnu": "^0.0.0",
    "skribble-css-linux-x64-musl": "^0.0.0",
    "skribble-css-wasm": "^0.0.0",
    "skribble-css-win32-arm64-msvc": "^0.0.0",
    "skribble-css-win32-ia32-msvc": "^0.0.0",
    "skribble-css-win32-x64-msvc": "^0.0.0"
  },
  "engines": {
    "node": ">= 14"
  },
  "publishConfig": {
    "access": "public"
  },
  "monots": {
    "entrypoints": [
      "index.ts",
      "client.ts"
    ],
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node"
          ],
          "noEmit": true
        }
      },
      "test": {}
    },
    "mode": "library",
    "addExportsToEntrypoints": true,
    "externalModules": []
  },
  "napi": {
    "name": "skribble_css",
    "triples": {
      "additional": [
        "aarch64-apple-darwin",
        "aarch64-linux-android",
        "aarch64-unknown-linux-gnu",
        "aarch64-unknown-linux-musl",
        "aarch64-pc-windows-msvc",
        "armv7-unknown-linux-gnueabihf",
        "x86_64-unknown-linux-musl",
        "x86_64-unknown-freebsd",
        "i686-pc-windows-msvc",
        "armv7-linux-androideabi"
      ]
    }
  }
}
