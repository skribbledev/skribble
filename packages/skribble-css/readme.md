# skribble-css

> A better way to do typed css
