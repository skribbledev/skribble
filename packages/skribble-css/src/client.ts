export * from '@skribble-css/client';
