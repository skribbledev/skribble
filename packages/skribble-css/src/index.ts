export * from './skribble-css.js';
