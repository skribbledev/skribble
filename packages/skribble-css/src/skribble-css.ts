/**
 * Add your inline documentation
 */
export function skribbleCss() {
  return '';
}
