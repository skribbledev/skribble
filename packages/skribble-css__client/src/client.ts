import type { SkribbleCss, WithCustomClassName } from '@skribble-css/types';

import { overrides } from './overrides';

let shouldCache = true;

const cache = new Map<string, string | WithCustomClassName<SkribbleCss>>();

/**
 * Determine whether class names should be cached. By default caching is turned
 * on. Set to `false` to turn it off.
 *
 * This can cause a speed up since every time a class name is accessed it
 * creates a new proxy.
 *
 * I haven't tested this, but potentially for a sufficiently large code base the
 * cache could become large enough to reduce performance. A threshold would be
 * useful.
 */
export function setCaching(should: boolean) {
  shouldCache = should;
}

/**
 * All the classNames which have been defined in the codebase.
 *
 * ```ts
 * import { c } from 'skribble-css/client';
 *
 * const Component = () => {
 *   return <div className={c.text.$lg} />;
 * }
 * ```
 */
export const c = createProxyClassNames();

/**
 * A utility function for constructing className strings conditionally. It also
 * has the added benefit of solving the specificity problem for the skribble css
 * classes.
 *
 * Typically with libraries like `tailwind` and `unocss` the style that is
 * rendered depends on the setup of the css file. This means that overriding a
 * style is determined outside of the JS file.
 *
 * With Skribble, this problem is addressed by checking the classnames and
 * removing any that use the same styles.
 */
export function cx(...values: ClassValue[]): string {
  const strings: string[] = [];

  for (const value of values) {
    if (!value) {
      continue;
    }

    const extracted = getCxValue(value);
    extracted && strings.push(extracted);
  }

  return deduplicateClassNames(strings).join(' ');
}

export type ClassValue =
  | ClassArray
  | ClassDictionary
  | string
  | number
  | null
  | boolean
  | undefined;

export interface ClassDictionary {
  [id: string]: any;
}

export interface ClassArray extends Array<ClassValue> {}

function generateArgsList(args: unknown[]) {
  return `[${args
    .filter((arg): arg is string => typeof arg === 'string')
    .map((arg) => arg.trim())
    .join(':')}]`;
}

function initialProxy(...args: unknown[]) {
  return generateArgsList(args);
}

/**
 * Create the class name proxy.
 */
function createProxyClassNames(v?: Set<string>): WithCustomClassName<SkribbleCss> {
  // const values: Set<string> = new Set(v ? [...v] : []);

  const proxy = new Proxy(initialProxy, {
    apply: (_, __, args) => {
      const values: Set<string> = v ? v : new Set();
      const props = [...values];
      const last = props[values.size - 1];

      const id = `${props.join('.')}--${args.join('.')}`;

      if (shouldCache) {
        const cachedValue = cache.get(id);

        if (typeof cachedValue === 'string') {
          return cachedValue;
        }
      }

      let className = '';

      if (last?.startsWith('$')) {
        throw new TypeError(`'${props.join('.')}' is not a function.`);
      }

      if (args.length === 0) {
        throw new TypeError(`'${props.join('.')}' must have at least one argument.`);
      }

      if (props.length > 0) {
        className += `${props.join(':')}::`;
      }

      className += generateArgsList(args);

      if (shouldCache) {
        cache.set(id, className);
      }

      return className;
    },

    get: (_, prop) => {
      const values: Set<string> = v ? v : new Set();

      if (typeof prop !== 'string') {
        return String.prototype[prop as keyof string];
      }

      if (typeof String.prototype[prop as keyof string] === 'function') {
        return (String.prototype[prop as keyof string] as any).bind('');
      }

      const id = [...values, prop].join('.');

      if (shouldCache) {
        const cachedValue = cache.get(id);

        if (cachedValue) {
          return cachedValue;
        }
      }

      // This is a value prop which completes the chain.
      if (prop.startsWith('$')) {
        return values.size > 0 ? `${[...values].join(':')}::${prop}` : prop;
      }

      const proxy = createProxyClassNames(new Set([...values, prop]));

      if (shouldCache) {
        cache.set(id, proxy);
      }

      return proxy;
    },

    set: (object, prop) => {
      throw new TypeError(
        `Cannot assign to read only property '${String(prop)}' of object '#${object}'`,
      );
    },
  });

  return proxy as any;
}

function getCxValue(value: NonNullable<ClassValue>): string {
  const output: string[] = [];

  if (typeof value === 'string' || typeof value === 'number') {
    return `${value}`;
  }

  if (Array.isArray(value)) {
    for (const item of value) {
      if (!item) {
        continue;
      }

      const extracted = getCxValue(item);
      extracted && output.push(extracted);
    }

    return output.join(' ');
  }

  if (typeof value === 'object') {
    for (const [key, val] of Object.entries(value)) {
      val && output.push(key);
    }

    return output.join(' ');
  }

  return '';
}

function deduplicateClassNames(classNames: readonly string[]) {
  const newClassNames: string[] = [];
  const removalStatus = [...classNames.map(() => false)];
  // const shorthandTarget: Record<string, number> = {};
  const valueTarget: Record<string, number> = {};
  const argumentsTarget: Record<string, number> = {};

  for (const [index, className] of classNames.entries()) {
    // Split by the divider `::`
    let [prefix = '', value = '', ...rest] = className.split('::');

    if (rest.length > 0 || !prefix) {
      continue;
    }

    if (!value) {
      value = prefix;
      prefix = '';
    }

    if (
      // Value Target - `$px`
      value.startsWith('$') ||
      //Argument Target - `[10px]`
      isArgument(value, 1)
    ) {
      const indexToRemove = valueTarget[prefix];

      if (typeof indexToRemove === 'number') {
        removalStatus[indexToRemove] = true;
      }

      const split = prefix.split(':');
      const modifiers = split.slice(0, -2);
      const alias = prefix && split.slice(-1)[0];
      const children = alias ? overrides.get(alias) : [];

      for (const child of children) {
        const key = [...modifiers, child].join(':');
        const childIndexToRemove = valueTarget[key];

        if (typeof childIndexToRemove === 'number') {
          removalStatus[childIndexToRemove] = true;
        }
      }

      valueTarget[prefix] = index;
    }

    // Arguments Target - `[padding:10px]`
    if (isArgument(value, 2)) {
      const [key] = value.split(':');
      prefix += key;
      const indexToRemove = argumentsTarget[prefix];

      if (typeof indexToRemove === 'number') {
        removalStatus[indexToRemove] = true;
      }

      argumentsTarget[prefix] = index;
    }
  }

  for (const [index, className] of classNames.entries()) {
    if (removalStatus[index]) {
      continue;
    }

    newClassNames.push(className);
  }

  return newClassNames;
}

function isArgument(value: string, length: 1 | 2): boolean {
  return value.startsWith('[') && value.endsWith(']') && value.split(':').length === length;
}
