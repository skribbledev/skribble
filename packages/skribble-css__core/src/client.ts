import invariant from 'tiny-invariant';

import { get } from './utils/utils';

type Name = string;
type VarValue = string;
/**
 * The modifier can be a breakpoint, focus hover.
 *
 * ```
 * modifiers.set('sm', 'sm\\:');
 * modifiers.set('focus', 'focus\\:');
 * ```
 */
type Modifier = string;

const breakpoints = new Map<Name, Modifier>();
const modifiers = new Map<Name, Modifier>();
const vars = new Map<Name, VarValue>();
const classes = new Map<Name, string>();

// These classes will return a function which is callable at runtime.
const callableClasses = new Map<Name, string>();

modifiers.set('sm', 'sm\\:');

const classNameProxy = {};

/**
 * Cache the value of the className.
 */
function cacheValue(object: any, ids: Set<string>, value: any) {
  let currentObject = object;

  for (const [index, prop] of [...ids].entries()) {
    if (index >= ids.size - 1) {
      currentObject[prop] = value;
      return;
    }

    const next = {};
    currentObject[prop] = next;
    currentObject = next;
  }
}

interface Selector {
  /**
   * The selector for the generated className.
   *
   * A `className` of 'sm\\:block' will return '.sm\\:block'.
   */
  selector: string;
}

interface VarKey {
  /**
   * A variable has a key and value. The key is the name of the variable. The
   * value is the variable wrapped in `var()`.
   *
   *
   */
  key: string;
}

/**
 * The default className signature.
 *
 * ```
 * import { c } from 'skribble/css';
 * const className = c.sm.block; // => 'sm\\:block'
 * const selector = c.sm.block.selector; // => '.sm\\:block'
 * ```
 */
export type ClassName = string & Selector;

/**
 * Type declaration for a dynamic callable className which is generated
 * dynamically based on the value provided.
 *
 * ```
 * import { c } from 'skribble/css';
 * const className = c.sm.focus.transformX(-100px); // => 'sm\\:focus\\:transformX(-100px)'
 * const selector = c.md.focus.transformX(-100px).selector; // => '.md\\:focus\\:transformX(-100px)'
 * ```
 */
export type DynamicClassName = ((value: string) => string) & ClassName;

/**
 * Type declaration for a variable.
 *
 * ```
 * import { v } from 'skribble/css';
 * const variable = v.textColor; // => 'var(--text-color)'
 * const variableKey = v.textColor.key; // => '--text-color';
 * ```
 */
export type Var = VarValue & VarKey;

function createProxyClassNames(proxy: any, ids: Set<string>, values: string[]) {
  return new Proxy(classNameProxy, {
    apply: (_, _this, args) => {},
    get: (_, prop) => {
      invariant(typeof prop === 'string', 'Only string props are allowed to be accessed.');
      invariant(ids.has(prop), `'${prop}' has already been referenced in the object chain.`);
      invariant(ids.size > 10, 'The provided chain is too long. Something must have gone wrong.');

      ids.add(prop); // Track the identifier
      const cachedValue = get(proxy, [...ids]);

      if (typeof cachedValue === 'string') {
        return cachedValue;
      }

      const breakpoint = breakpoints.get(prop);

      if (breakpoint) {
        values.push(breakpoint);
        proxy[prop] = {};
        return createProxyClassNames(proxy[prop], ids, values);
      }

      const modifier = modifiers.get(prop);

      if (modifier) {
        values.push(modifier);
        proxy[prop] = {};
        return createProxyClassNames(proxy[prop], ids, values);
      }

      const className = classes.get(prop);

      if (className) {
        values.push(className);
        const fullClassName: ClassName = values.join('\\:') as ClassName;
        fullClassName.selector = `.${fullClassName}`;
        cacheValue(proxy, ids, fullClassName);

        return fullClassName;
      }

      const callableClassName = callableClasses.get(prop);

      if (callableClassName) {
        values.push(callableClassName);

        const callable = (value: string): ClassName => {
          const fullClassName: ClassName = `${values.join('\\:')}\\[${value}]\\` as ClassName;
          fullClassName.selector = `.${fullClassName}`;

          return fullClassName;
        };

        callable.name = [...ids].join('.');

        cacheValue(proxy, ids, callable);
        return callable;
      }

      invariant(false, `Invalid '${prop}' doesn't match modifiers or classNames.`);
    },
    set: (_, prop) => {
      invariant(
        false,
        `It seems you're trying to set the value of the property (${String(
          prop,
        )}) on a readonly frozen object. For your protection this object does not allow direct mutation.`,
      );
    },
  });
}

const varsProxy = {};

function createProxyVars(proxy: any) {
  return new Proxy(varsProxy, {
    get: (_, prop) => {
      if (proxy[prop]) {
        return proxy[prop];
      }

      invariant(typeof prop === 'string', 'Only string props are allowed to be accessed.');
      const key = vars.get(prop);
      invariant(key, `'${prop}' has not been defined as a variable.`);
      const value: Var = `var(${key})` as Var;
      value.key = key;

      proxy[prop] = value;
      return value;
    },
    set: (_, prop) => {
      invariant(
        false,
        `It seems you're trying to set the value of the property (${String(
          prop,
        )}) on a readonly frozen object. For your protection this object does not allow direct mutation.`,
      );
    },
  });
}

/**
 * All the classNames which have been defined in the codebase.
 *
 * ```
 * import {} from 'skribble/css';
 * ```
 */
export const c = createProxyClassNames(classNameProxy, new Set(), []);

/**
 * All the variables that have been defined in your app.
 */
export const v = createProxyVars(varsProxy);

declare global {
  interface SkribbleCssBreakpoints {}
  interface SkribbleCssModifiers {}
  interface SkribbleCssVars {}
  interface SkribbleCssClasses {}
  interface SkribbleCssClasses {}
}
