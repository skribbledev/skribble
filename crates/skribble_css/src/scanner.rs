pub mod class_name;
pub mod class_name_collector;
mod class_name_order;

#[cfg(test)]
mod tests {
  use crate::test_utils::{snapshot_selector, snapshot_selectors, test_no_selector};

  snapshot_selector!(generate_class_name:
    "import { c } from 'skribble-css/client';
    const d = c.md.focus.$block;");

  snapshot_selectors!(generate_custom_class:
    "import { c } from 'skribble-css/client';
    c('padding', '10px');
    c.md('padding', '6px');
    c.sm('padding', '16px');"
  );

  snapshot_selectors!(generate_class_names:
    "import { c } from 'skribble-css/client';
    const d = c.md.focus.$block;
    const e = c.sm.focus.$block;
    const f = c.lg.focus.$block;");

  snapshot_selectors!(dedupe_class_names:
    "import { c } from 'skribble-css/client';
    const a = c.md.focus.$block;
    const b = c.md.focus.$block;
    const d = c.sm.focus.$block;
    const e = c.sm.focus.$block;
    const f = c.lg.focus.$block;
    const g = c.lg.focus.$block;");

  snapshot_selector!(import_as: "\
    import { c as b } from 'skribble-css/client';
    b.sm.focus.$grid
    "
  );

  snapshot_selector!(inside_scope: "\
    import { c as b } from 'skribble-css/client';
    {
      b.lg.focus.$block;
    }"
  );

  test_no_selector!(no_shorthand_or_style_name:
    "import { c } from 'skribble-css/client';
    c.sm.focus.grid;"
  );

  snapshot_selector!(inside_function_scope:
    "import { c } from 'skribble-css/client';

    function fun () {
      c.lg.focus.$block;
    }");

  snapshot_selector!(default_parameter:
    "import { c } from 'skribble-css/client';

    function awesome(a = c.lg.focus.$block) {
     //
    }"
  );

  test_no_selector!(ignore_external_identifiers_in_callable_fns:
  "import { c } from 'skribble-css/client';
    // This is ignored.
    import { ehh } from 'ehh';

    c.md.p(ehh);"
  );

  snapshot_selector!(identifiers_can_be_evaluated:
  "import { c } from 'skribble-css/client';

    const ehh  = '10px';

    c.md.p(ehh);",
    ignore
  );

  snapshot_selectors!(generate_from_arrays:
    "import { c } from 'skribble-css/client';
    const array = [c.md.focus.$block,c.sm.focus.$block,c.lg.focus.$block];"
  );

  snapshot_selectors!(generate_from_objects: "\
    import { c } from 'skribble-css/client';
    const obj = {
      a: c.md.focus.$block,
      b: c.sm.focus.$block,
      c: c.lg.focus.$block
    };"
  );

  snapshot_selectors!(generate_from_object_keys: "\
    import { c } from 'skribble-css/client';

    const obj = {
      [c.md.focus.$block]: true,
      [c.sm.focus.$block]: true,
      [c.lg.focus.$block]: false,
    };"
  );

  snapshot_selector!(support_literals_in_expression: "
    import { c } from 'skribble-css/client';
    c['md'].p(1n);"
  );

  snapshot_selector!(jsx: "\
    import { c } from 'skribble-css/client';

    const Component = () => {
      return <div className={c.sm.m.$1} />;
    };"
  );

  snapshot_selector!(function_args: "\
    import clsx from 'clsx';
    import { c } from 'skribble-css/client';
    clsx(c.sm.m.$1)
    "
  );

  snapshot_selector!(jsx_prop_within_function: "\
    // import clsx from 'clsx';
    import { c } from 'skribble-css/client';

    const Component = () => {
      return <div className={clsx(c.sm.m.$1)} />;
    };"
  );

  snapshot_selector!(jsx_array_prop: "\
    import Other from 'library';
    import { c } from 'skribble-css/client';

    const Component = () => {
      return <Other style={[c.sm.m.$1]} />;
    };"
  );

  snapshot_selector!(jsx_array_object: "\
    import Other from 'library';
    import { c } from 'skribble-css/client';

    const Component = () => {
      return <Other style={{ a: c.sm.m.$1 }} />;
    };"
  );

  snapshot_selector!(jsx_array_object_computed: "\
    import Other from 'library';
    import { c } from 'skribble-css/client';

    const Component = () => {
      return <Other style={{ [c.sm.m.$1]: true }} />;
    };"
  );

  snapshot_selector!(supports_leading_underscore: "\
    import Other from 'library';
    import { c } from 'skribble-css/client';

    c.sm.p.$1;"
  );

  snapshot_selector!(support_colors: "\
    import { c } from 'skribble-css/client';

    c.text.$primary;"
  );

  snapshot_selector!(escape_custom_css_value: "\
    import { c } from 'skribble-css/client';

    c.focus.text('rgb(0,0,0)');"
  );

  // Scoping should mean that the variable is overwritten here.
  test_no_selector!(scope_can_override_import: "\
  import { c } from 'skribble-css/client';
  {
    const c = { md: 'md' };
    const d = c.md.c
  }", ignore, should_panic);
}
