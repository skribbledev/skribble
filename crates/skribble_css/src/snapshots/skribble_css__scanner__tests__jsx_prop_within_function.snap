---
source: crates/skribble_css/src/scanner.rs
assertion_line: 128
expression: "crate::test_utils::get_selector(&crate::test_utils::create_config(None).unwrap(),\n                                \"\\\n    // import clsx from 'clsx';\n    import { c } from 'skribble-css/client';\n\n    const Component = () => {\n      return <div className={clsx(c.sm.m.$1)} />;\n    };\")"

---
.sm\:m\:\:\$1
